doc/ts_setup.3
.\" Copyright (c) 2017, Martin Kepplinger <martink@posteo.de>
.\"
.\" %%%LICENSE_START(GPLv2+_DOC_FULL)
.\" This is free documentation; you can redistribute it and/or
.\" modify it under the terms of the GNU General Public License as
.\" published by the Free Software Foundation; either version 2 of
.\" the License, or (at your option) any later version.
.\"
.\" The GNU General Public License's references to "object code"
.\" and "executables" are to be interpreted as the output of any
.\" document formatting or typesetting system, including
.\" intermediate and printed output.
.\"
.\" This manual is distributed in the hope that it will be useful,
.\" but WITHOUT ANY WARRANTY; without even the implied warranty of
.\" MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
.\" GNU General Public License for more details.
.\"
.\" You should have received a copy of the GNU General Public
.\" License along with this manual; if not, see
.\" <http://www.gnu.org/licenses/>.
.\" %%%LICENSE_END
.\"
.TH TS_SETUP 3  "" "" "tslib"
.SH NAME
ts_setup \- find, open and configure a touch screen input device
.SH SYNOPSIS
.nf
.B #include <tslib.h>
.sp
.BI "struct tsdev *ts_setup(const char *" dev_name ", int " nonblock ");"
.sp
.fi

.SH DESCRIPTION
.BR ts_setup ()
will try to find, open and configure the touch screen device.
On success it allocates memory for a new touch screen device, opens and links the input event device to it
and loads and initialises all modules configured in the tslib configuration file
.BR TSLIB_CONFFILE.

If \fBdev_name\fR is \fBNULL\fR
.BR ts_setup ()
will use the device defined by the \fBTSLIB_TSDEVICE\fR environment variable.

If \fBTSLIB_TSDEVICE\fR is not defined
.BR ts_setup ()
will try to use following devices:
.RS
/dev/input/ts,
.br
/dev/input/touchscreen,
.br
/dev/touchscreen/ucb1x00.
.RE
After that we ts_setup() scans /dev/input/event* for the first device with property INPUT_PROP_DIRECT.

.SH RETURN VALUE
A pointer to the opened
.BI "struct tsdev"
is returned.

.SH SEE ALSO
.BR ts_open (3),
.BR ts_close (3),
.BR ts_read (3),
.BR ts_config (3),
.BR ts.conf (5)


NEWS
tslib 1.14 - tbd
================================

This release includes libts version 0.8.1 and the following changes:
* The following modules are now disabled in the default configuration run:
 * `module_raw arctic2`
 * `module_raw corgi`
 * `module_raw collie`
 * `module_raw dmc`
 * `module_raw h3600`
 * `module_raw mk712`
 * `module_raw ucb1x00`
 Users of those modules must enable them, like `./configure --enable-arctic2`
* Automatic touchscreen device detection for Linux (if not defined by the user)
* New filter plugin "lowpass"

This release includes the following bugfixes:
* 93ac81d autoconf: disable dependency tracking by default (fixes out-of-source builds)
* f4cd766 waveshare-raw: fix snprintf error in finding a device in ts_read()
* 57d9081 debounce: drop samples irrespective of pressure value
* 5822386 waveshare: fix resource leak


tslib 1.13 - released 2017-09-14
================================

This release includes libts version 0.8.0 and the following changes:
* Added ts_open_restricted() and ts_close_restricted() API
* Added TSLIB_MT_VALID bit definition for ts_sample_mt's valid flag
* `module_raw galax` is now disabled in the default configuration.
  Please use `module_raw input` instead.

This release includes the following bugfixes:
* 18dfd8b remove build dependency on pkg-config
* f10104e configure.ac: add GNU/kFreeBSD to build config
* b8594cc input-raw: Don't require to have the EV_KEY event type
* 1ad02d4 input-raw.c: support MT-only devices in ts_read()
* 00c69d6 fix static building for galax and cy8mrln-palmpre
* 5a8b080 Update fbutils-linux.c to fix fillrect only drawing first y line


tslib 1.12 - released 2017-08-29
================================

This release includes libts version 0.7.2 and the following changes:
* Generate a tracking ID value for multitouch type A devices
* Add experimental `--with-sdl2` configuration option for `ts_calibrate` and `ts_test_mt`

This release includes the following bugfixes:
* 74f527e configure.ac: fix help text by defining only one default configuration


tslib 1.11 - released 2017-06-06
================================

This release includes libts version 0.7.1 and the following changes:
* The graphical tools don't show TSLIB anymore, but Touchscreen instead.
* The graphical tools support rotating the screen 90, 180 and 270 degrees
* man page updates

This release includes the following bugfixes:
* ff8c7d9 use dlfcn.h dynamic loading function only if libdl is available
* d39d358 input-raw: fix multitouch devices without BTN_TOUCH


tslib 1.10 - released 2017-05-12
================================

This release includes libts version 0.7.0 and the following changes:
* new `ts_libversion()` API and TSLIB_VERSION_MT definition
  This offers seamless integration for apps that don't know tslib's version
* new `module_raw dmc_dus3000` for DMC DUS Series touch screen via UART
* Fixed and improved statically linked building

This release includes the following bugfixes:
* c03f7c9 tools: ts_uinput_start.sh: fix usage of ts_uinput -d -v
* 47e7ab5 input-raw: Fix wrong first slot values by using calloc instead of malloc
* 7c5b4f8 plugins: filters: check for read_mt() availability
* e684df5 call ops->fini() only when implemented by the plugin
* 0ca9580 ts_close: free tslib_module_info when fini() is not implemented
* 5974266 median: read_mt: mark all-zero samples invalid


tslib 1.9 - released 2017-04-21
===============================

This release includes libts version 0.6.0 and the following changes:
* new `module iir` infinite impulse response filter
* `module_raw galax` is deprecated. Users should switch to `module_raw input`
* multitouch for (at least) eGalax 210 (`module_raw input` and `ts_read_mt()`)
* new `ts_verify` API test program
* `ts_uinput -d -v` prints the new /dev/input/eventX device node before returning
* experimental support for GNU / Hurd

This release includes the following bugfixes:
* fix support for protocol type A devices (for example Sitronix ST1232)
* fix the `skip` filter's read_mt() multitouch operation
* 338d2e4 input-raw: add MT protocol type A support to ts_read()
* e47df79 input-raw: fix pressure capable multitouch devices
* dbb8f4c ts_uinput: correctly return in case forking fails
* 73773ac fbutils-linux: fix framebuffer errors on 64 bit systems
* 4d63bfa input-raw: fix ts_read_mt for non-MT devices without pressure


tslib 1.8 - released 2017-03-21
===============================

tslib 1.8 is backwards compatible. Changes since last release include:
* experimental support for building on Windows (using MSYS2)
* fixed ts_read_mt() in non-blocking operation when trying to read more than
  1 sample, see bugfixes
* added workaround for strangely behaving eGalax model 210 devices, see bugfixes
* bugfixes
  - 388d3da input-raw: ts_read_mt: fix returning actually read samples
  - c7f0966 debounce: use signed time values
  - 9a9b0c9 pthres: read_mt: fix internal pressure state handling
  - fa10e4c galax: version 210: use ABS_MT_DISTANCE to find pressure 0


tslib 1.7 - released 2017-03-13
===============================

Changes since last release:
* ts_read_mt() now consistently returns errno set by read() when using
  module_raw input
* extended ts_print_mt to test more of our API
* ts_print_mt --raw replaces the ts_print_raw_mt program
* basic support for the Haiku OS
* added a release helper script to simplify the releasing process
* add proper AUTHORS and THANKS files, as well as .mailmap
* bugfixes
  - 6ea643e ts_test_mt: fix idev commandline option
  - 9617c7d dmc-raw: fix errno return code to be negative
  - dc2fdaf input-raw: remove errror message on each negative return from read()
  - 75b5f40 input-raw: return read() errno on error.
  - de5e261 skip: fix buffer memory freeing
  - da5477e median: fix ts_read_mt() pen down event


tslib 1.6 - released 2017-03-03
===============================

Changes since the last release:
* ABI addition: libts.so.0 (ensuring backwards compatibility)

src/ts_setup.c
/*
 *  tslib/src/ts_setup.c
 *
 *  Copyright (C) 2017 Piotr Figlarek
 *
 * This file is placed under the LGPL.  Please see the file
 * COPYING for more details.
 *
 *
 * Find, open and configure a touchscreen device.
 */

#include "tslib.h"
#include "tslib-private.h"
#include <stdlib.h>
#include <stdio.h>
#include <errno.h>
#include <string.h>

#if defined (__linux__) 
#include <dirent.h>
#include <stdio.h>
#include <string.h>
#include <linux/input.h>
#include <sys/types.h>
#include <sys/stat.h>
#include <fcntl.h>
#include <unistd.h>

#define DEV_INPUT_EVENT "/dev/input"
#define EVENT_DEV_NAME "event"

#define DIV_ROUND_UP(n, d) (((n) + (d) - 1) / (d))
#define BIT(nr)                 (1UL << (nr))
#define BIT_MASK(nr)            (1UL << ((nr) % BITS_PER_LONG))
#define BIT_WORD(nr)            ((nr) / BITS_PER_LONG)
#define BITS_PER_BYTE           8
#define BITS_PER_LONG           (sizeof(long) * BITS_PER_BYTE)
#define BITS_TO_LONGS(nr)       DIV_ROUND_UP(nr, BITS_PER_BYTE * sizeof(long))

static int is_event_device(const struct dirent *dir) {
	return strncmp(EVENT_DEV_NAME, dir->d_name, 5) == 0;
}

static char* scan_devices(void)
{
	struct dirent **namelist;
	int i, ndev;
	char *filename = NULL;
	int have_touchscreen = 0;
	long propbit[BITS_TO_LONGS(INPUT_PROP_MAX)] = {0};

#ifdef DEBUG
	printf("scanning for devices in %s\n", DEV_INPUT_EVENT);
#endif

	ndev = scandir(DEV_INPUT_EVENT, &namelist, is_event_device, alphasort);
	if (ndev <= 0)
		return NULL;

	for (i = 0; i < ndev; i++) {
		char fname[64];
		int fd = -1;

		snprintf(fname, sizeof(fname),
			 "%s/%s", DEV_INPUT_EVENT, namelist[i]->d_name);
		fd = open(fname, O_RDONLY);
		if (fd < 0)
			continue;

		if ((ioctl(fd, EVIOCGPROP(sizeof(propbit)), propbit) < 0) ||
			!(propbit[BIT_WORD(INPUT_PROP_DIRECT)] &
				  BIT_MASK(INPUT_PROP_DIRECT)) ) {
			continue;
		} else {
			have_touchscreen = 1;
		}

		close(fd);
		free(namelist[i]);

                if (have_touchscreen) {
			filename = malloc(strlen(DEV_INPUT_EVENT) +
					  strlen(EVENT_DEV_NAME) +
					  3);
			if (!filename)
				return NULL;

	                sprintf(filename, "%s/%s%d",
				DEV_INPUT_EVENT, EVENT_DEV_NAME,
				i);
		}

		return filename;
	}

	return NULL;
}

#endif /* __linux__ */

static const char * const ts_name_default[] = {
		"/dev/input/ts",
		"/dev/input/touchscreen",
		"/dev/touchscreen/ucb1x00",
		NULL
};

struct tsdev *ts_setup(const char *dev_name, int nonblock)
{
	const char * const *defname;
	struct tsdev *ts = NULL;
#if defined (__linux__)
	char *fname = NULL;
#endif /* __linux__ */

	dev_name = dev_name ? dev_name : getenv("TSLIB_TSDEVICE");

	if (dev_name != NULL) {
		ts = ts_open(dev_name, nonblock);
	} else {
		defname = &ts_name_default[0];
		while (*defname != NULL) {
			ts = ts_open(*defname, nonblock);
			if (ts != NULL)
				break;

			++defname;
		}
	}

#if defined (__linux__)
	if (!ts) {
		fname = scan_devices();
		if (!fname)
			return NULL;

		ts = ts_open(fname, nonblock);
		free(fname);
	}
#endif /* __linux__ */

	/* if detected try to configure it */
	if (ts && ts_config(ts) != 0) {
		ts_error("ts_config: %s\n", strerror(errno));
		ts_close(ts);
		return NULL;
	}

	return ts;
}
